src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java

```diff
@@ -609,7 +609,7 @@ protected void connect(SocketAddress remote, int millis) throws IOException {
         } catch (IOException ioe) {
             close();
             if (ioe instanceof InterruptedIOException) {
-                throw ioe;
+                throw new SocketException("Closed by interrupt");
```

Comment:
SocketTimeoutException is InterruptedIOException so need to distinguish this from the case where a virtual thread is interrupt during connect. The change for NioSocketImpl needs to be like this (once exception handle in switch comes then we can improve this).

```
             }
         } catch (IOException ioe) {
             close();
-            if (ioe instanceof InterruptedIOException) {
+            if (ioe instanceof SocketTimeoutException) {
                 throw ioe;
+            } else if (ioe instanceof InterruptedIOException) {
+                assert Thread.currentThread().isVirtual();
+                throw new SocketException("Closed by interrupt");
             } else {
                 throw SocketExceptions.of(ioe, isa);
             }
```